Add n31363 TxGraph notes (#786)
bitcoin-core-review-club/website: _posts/2025-02-05-#31363.md

commit:
---

## Notes

### General Cluster Mempool

Skip this section if you are just interested in reviewing the code and don't need to hear the pitch for why.

> The primary goal of [Cluster Mempool](https://github.com/bitcoin/bitcoin/issues/30289) is to enable the mempool to more

Reviewer: Probably worth linking to the optech write-up here too, it's good! https://bitcoinops.org/en/topics/cluster-mempool/

censorship problems.

- "Why is cluster-tracking the solution for incentive compatibility?"
  - [This Delving post](https://delvingbitcoin.org/t/mempool-incentive-compatibility/553) provides a great explanation of why the question of incentive
    compatibility is complicated and why tracking clusters is necessary for a complete solution.
  - One of the benefits of Topologically Restricted Until Confirmation is to have an effective cluster limit within
the existing framework of ancestor/descendant limits (limit of cluster count 2 is the equivalent of limiting ancestor
and descendant counts to 2). However, a limit of 2 is quite prohibitive.
  - We talked about clusters in previous review club meetings about [auto-bumping unconfirmed UTXOs in coin selection](/26152)
    (which added `CTxMmemPool::GatherClusters`) and [v3 transactions](/25038).

Mempool transactions and their dependency relationships can naturally be represented as a directed graph where nodes are
transactions and edges represent UTXO spending relationships. A transaction's _cluster_ is the part of the graph that it

> is connected to.

Reviewer: I think there's some potential for naming ambiguity here, that might be worth touching on? If my understanding is correct:
A cluster is the set of all nodes that are indirectly connected to each other through UTXO spending relationships. I think that generally also falls under the definition of a graph? So "A transaction's _cluster_ is the part of the graph that it is connected to." can perhaps be a bit confusing? I think a cluster is a graph, and `TxGraph` represents all the graphs/clusters that exist in the mempool?

I don't think this OP has to define the terms, but I think highlighting this might help avoid confusion?

> Currently (on master or as of v28), Bitcoin Core's mempool does not track clusters, but does track ancestor and

Reviewer: ```suggestion
On master, Bitcoin Core's mempool does not track clusters, but it does track ancestor and
```

If you have more time to spend on background or need more convincing, here is additional OPTIONAL reading material.

- [BIP 431](https://github.com/bitcoin/bips/blob/master/bip-0431.mediawiki) provides an overview of why RBF pinning
  is problematic and some of the other pieces that come together to address the issue.
- [This Delving post](https://delvingbitcoin.org/t/cluster-mempool-definitions-theory/202) provides an overview of the
  algorithms used in cluster linearization. 
- [This Delving post](https://delvingbitcoin.org/t/cluster-mempool-rbf-thoughts/156) discusses how to evaluate
  replacements using feerate diagrams, which are enabled with cluster mempool.

### Background on Mempool Architecture and Interfaces

Skip this section if you are already familiar with current mempool code.

"The Mempool" usually refers to the class [`CTxMemPool`](https://github.com/bitcoin/bitcoin/blob/8fa10edcd1706a1f0dc9d8c3adbc8efa3c7755bf/src/txmempool.h#L303).
This class holds a map of `CTxMemPoolEntry`s and other supporting data structures. It is responsible for keeping itself
internally consistent, and much of the code is dedicated to dealing with dependency relationships.

> The logic surrounding mempool, where much of the fun code lives, includes:

Reviewer: ```suggestion
Interesting mempool implementation code includes:
```

The responsibility of keeping this graph up-to-date and within its specified limits largely belongs to `CTxMemPool`, but
occasionally bleeds out to other areas. For example, in mempool validation, `MemPoolAccept` queries `CTxMemPool` for the
ancestors of a transaction in order to:

- Calculate whether this transaction would exceed the ancestor/descendant limits. However, sometimes, when evaluating a
  replacement, we apply an [RBF Carve Out](https://github.com/bitcoin/bitcoin/blob/8fa10edcd1706a1f0dc9d8c3adbc8efa3c7755bf/src/validation.cpp#L950-L984),
literally increasing the limit by 1, to avoid double-counting the replacement and to-be-replaced transactions. Yuck!
- Check whether a TRUC transaction meets its additional topology restrictions. However we first need to [aggregate the
  in-package and in-mempool ancestors](https://github.com/bitcoin/bitcoin/blob/8fa10edcd1706a1f0dc9d8c3adbc8efa3c7755bf/src/validation.cpp#L1551-L1558).
- When evaluating a package, the precise ancestor/descendant counts of each package transaction and connected mempool
  transactions are not calculated before submission. Instead, the aggregate package count and size are used. This

> approximation avoids complexity, but can result in overestimation for topologies beyond 1-parent-1-child.

Reviewer: > The responsibility of keeping this graph up-to-date and within its specified limits largely belongs to `CTxMemPool`

Are you talking about the "emergent" graph that exists on master, or the new TxGraph? If the former, perhaps this section might be a bit of information overload - would it make sense to leave it out?

amount of time you have to review it.

#### Basic TxGraph, Lazy and Oversized

This section includes code up to [txgraph: (optimization) avoid per-group vectors for clusters & dependencies](https://github.com/bitcoin-core-review-club/bitcoin/commit/5e55a584a77a8dfa357eaf016f177d72cb92a218).

A `TxGraph` has a vector of `Entry`s, where each entry is a single transaction, and a vector of `Cluster`s.  It also has
queues of "work" to do which may or may not need to be completed in order to perform certain tasks or answer external
queries. Some tasks have other tasks as prerequisites.

This is the order in which things have to be resolved:

1. `ApplyRemovals`: Apply all of the queued removals.
2. `SplitAll`: Split up clusters if they contain components that are no longer connected to each other.
3. `GroupClusters`: Calculate what clusters would need to be merged if dependencies are applied. However, these changes are not yet applied.
4. `ApplyDependencies`: Apply new dependencies.
5. Re-linearize clusters.

Much of the work is done lazily. For example, calling `RemoveTransaction` does not actually mean its etry is deleted, just that

> the index has been added to `m_to_remove`. However, txgraph cannot answer a call to `Exists` without first calling `ApplyRemovals`.

Reviewer: stylistic nit: s/txgraph/`TxGraph`/g

An **Oversized** `TxGraph` is one that has at least one cluster that violates the cluster count or size limits.
- A `TxGraph` only knows if it is oversized after it has completed steps 1-3. 
- When a `TxGraph` is oversized, mutation functions (like `RemoveTransaction` and `AddDependency`) and

>   `GetIndividualFeerate` can still be called, but inspection functions like `` are unavailable because we won't

Reviewer: todo: missing inspection functions

linearize clusters that are oversized.
- Additionally, step 4 will not proceed if the txgraph is oversized; step 5 is thus also impossible.

The `GroupClusters` function implements [union-find](https://en.wikipedia.org/wiki/Disjoint-set_data_structure), a
favorite amongst software engineering interviewers.

<!-- FIXME: Unlinked -->

#### Staging Support

This section includes code up to ["txgraph: (feature) add staging support"](https://github.com/bitcoin-core-review-club/bitcoin/commit/79647fd58dc7152eb1dc4e09cc56db845608352d).

This feature allows users of `TxGraph` to stage changes and then apply or discard them, much like git.

> Much of the `TxGraph` members are grouped together as a `ClusterSet`: ther vector of `Cluster`s, work queues, and
> counter. This refactor is done so that a `TxGraph` can have multiple **subgraphs** or **levels** represented as

Reviewer: s/ther/the, but I think the second part can just be removed - I think explaining the refactor is the main point here?
```suggestion
Much of the `TxGraph` members are grouped together as a `ClusterSet`. This refactor is done so that a `TxGraph` can have multiple **subgraphs** or **levels** represented as
```

`ClusterSet`s: main (level 0) and staged (level 1).

[Creating](https://github.com/bitcoin-core-review-club/bitcoin/commit/79647fd58dc7152eb1dc4e09cc56db845608352d#diff-c25ebcc6e4dc9c9bb21462f796fbb470cee3f0ff1e77bfc4033f4e8460f4482cR1464)
a staged subgraph requires first applying removals. The staged subgraph creates copies of information from the main graph.

<!-- FIXME: more on impl -->

#### Specialized Functionality

> This section includes the rest of #31363, but also #31444.

Reviewer: ```suggestion
This section includes the rest of [#31363](https://github.com/bitcoin/bitcoin/pull/31363), but also [#31444](https://github.com/bitcoin/bitcoin/pull/31444).
```

- `CountDistinctClusters` and `GetMainStagingDiagram` are used in RBF logic.
- `GetBlockBuilder` is used in block assembly.
- `GetWorstMainChunk` is used in mempool eviction.

It may be useful to look at the code surrounding mempool to understand why each of the functions are necessary.

<!-- #### Discretional Size-Limiting -->
<!-- This is everything up to Trim(), i.e. #31553 -->

## Questions

<!-- Basic TxGraph -->

1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?

1. What are the benefits of having this `TxGraph` within `CTxMemPool`, in your own words? Can you think of drawbacks?

1. What is the difference between `TxGraph` and `Cluster`?

1. If a txgraph is oversized, which functions are still guaranteed to return an accurate result, and which ones aren't?

1. Under what circumstances could `GetIndividualFeerate` and `GetChunkFeerate` return nothing (i.e. an empty `FeeFrac`?)
   Is it possible that one returns an answer and the other doesn't? How?

1. Why is `AddTransaction` annotated with `[[nodiscard]]`?

> 1. After calling `TxGraph::RemoveTransaction`, is the transaction removed transaction?

Reviewer: I think there's a grammar issue here, perhaps you meant `transaction removed immediately`?

1. Why can't `ApplyDependencies` continue if the txgraph is oversized?

1. Why isn't `Exists` a `const` function like one would expect?

<!-- Staging Support -->

> 1. What is the purpose of supporting staging? In what situation would we use it and does it warrant this complexity?

Reviewer: ```suggestion
1. What is the purpose of `TxGraph` supporting the staging of update operations? In what situation would we use it, and do you think the benefits warrant the extra code complexity?
```

> 1. What are the `TxGraph` functions analogous to `git add`, `git commit`, and `git stash`?

Reviewer: Hah, that's a fun kind of question.

Author: I figured git is our bread and butter :P
